vllm/v1/core/kv_cache_utils.py

```diff
@@ -1296,6 +1296,82 @@ def _report_kv_cache_config(
     )
 
 
+def _auto_fit_max_model_len(
+    vllm_config: VllmConfig,
+    kv_cache_specs: list[dict[str, KVCacheSpec]],
+    available_memory: list[int],
+) -> None:
+    """
+    When max_model_len is set to -1, this function estimates the largest
+    context length that can be supported with the available GPU memory.
+    It uses binary search to find the maximum length that fits across all
+    workers.
+
+    Args:
+        vllm_config: The global VllmConfig (will be modified in-place)
+        kv_cache_specs: List of dict[layer_name, KVCacheSpec] for each worker.
+            These should already be unified (via get_kv_cache_groups) to match
+            the actual allocation strategy.
+        available_memory: Memory available for KV cache in bytes for each
+            worker.
+    """
+    original_max = vllm_config.model_config.max_model_len
+
+    # Find the minimum estimated max_model_len across all workers
+    estimated_max_lens = []
+    for kv_cache_spec_one_worker, available_memory_one_worker in zip(
+        kv_cache_specs, available_memory
+    ):
+        if not kv_cache_spec_one_worker:
+            # Skip empty specs (attention-free models)
+            continue
+
+        estimated = estimate_max_model_len(
+            vllm_config, kv_cache_spec_one_worker, available_memory_one_worker
+        )
+        estimated_max_lens.append(estimated)
+
+    # Restore the original max_model_len in case estimate_max_model_len modified it
+    vllm_config.model_config.max_model_len = original_max
```

Comment:
not strictly related to this PR, but I would prefer `estimate_max_model_len` didn't modify things in place since it's supposed to estimate a value, maybe it would be better to copy the config.
We don't have to act here though.